common/src/main/java/org/apache/celeborn/common/network/ssl/SSLFactory.java

```diff
@@ -0,0 +1,407 @@
+
+    /**
+     * Sets the Keystore and Keystore password
+     *
+     * @param keyStore The key store file to use
+     * @param keyStorePassword The password for the key store
+     * @return The builder object
+     */
+    public Builder keyStore(File keyStore, String keyStorePassword) {
+      this.keyStore = keyStore;
+      this.keyStorePassword = keyStorePassword;
+      return this;
+    }
+
+    /**
+     * Sets the key password
+     *
+     * @param keyPassword The password for the private key in the key store
+     * @return The builder object
+     */
+    public Builder keyPassword(String keyPassword) {
+      this.keyPassword = keyPassword;
+      return this;
+    }
+
+    /**
+     * Sets the trust-store, trust-store password, whether to use a Reloading TrustStore, and the
+     * trust-store reload interval, if enabled
+     *
+     * @param trustStore The trust store file to use
+     * @param trustStorePassword The password for the trust store
+     * @param trustStoreReloadingEnabled Whether trust store reloading is enabled
+     * @param trustStoreReloadIntervalMs The interval at which to reload the trust store file
+     * @return The builder object
+     */
+    public Builder trustStore(
+        File trustStore,
+        String trustStorePassword,
+        boolean trustStoreReloadingEnabled,
+        int trustStoreReloadIntervalMs) {
+      this.trustStore = trustStore;
+      this.trustStorePassword = trustStorePassword;
+      this.trustStoreReloadingEnabled = trustStoreReloadingEnabled;
+      this.trustStoreReloadIntervalMs = trustStoreReloadIntervalMs;
+      return this;
+    }
+
+    /**
+     * Builds our {@link SSLFactory}
+     *
+     * @return The built {@link SSLFactory}
+     */
+    public SSLFactory build() {
+      return new SSLFactory(this);
+    }
+  }
+
+  /**
+   * Returns an initialized {@link SSLContext}
+   *
+   * @param requestedProtocol The requested protocol to use
+   * @param keyManagers The list of key managers to use
+   * @param trustManagers The list of trust managers to use
+   * @return The built {@link SSLContext}
+   * @throws GeneralSecurityException
+   */
+  private static SSLContext createSSLContext(
+      String requestedProtocol, KeyManager[] keyManagers, TrustManager[] trustManagers)
+      throws GeneralSecurityException {
+    SSLContext sslContext = SSLContext.getInstance(requestedProtocol);
+    sslContext.init(keyManagers, trustManagers, null);
+    return sslContext;
+  }
+
+  /**
+   * Creates a new {@link SSLEngine}. Note that currently client auth is not supported
+   *
+   * @param isClient Whether the engine is used in a client context
+   * @param allocator The {@link ByteBufAllocator to use}
+   * @return A valid {@link SSLEngine}.
+   */
+  public SSLEngine createSSLEngine(boolean isClient, ByteBufAllocator allocator) {
+    SSLEngine engine = createEngine(isClient, allocator);
+    engine.setUseClientMode(isClient);
+    engine.setWantClientAuth(true);
```

Comment:
Unlike in spark, where clients do not have ssl cert, here we allow for ssl client to present their certificate - when available.
This is mainly relevant when master and workers communicate, and not so much between applications and master/worker (not common for cert to be present for spark/flink/mr apps) - providing for master/worker to validate their identity.

See `SslConnectivitySuiteJ.testUntrustedClientCertFails` where this validation failure is tested.

Reply:
> Unlike in spark, where clients do not have ssl cert, here we allow for ssl client to present their certificate - when available.

Would be great to do this on Spark too. I had a PR to do this that I never got around to submitting as part of other feature tests, but I wasn't sure if it would have value in isolation beyond the other changes I had in mind (those proved infeasible).

Reply:
The api expects an empty array and not null to be returned.
As part of testing various combinations, this surfaced, yes.

We need something similar to ssl connectivity test in Spark as well.